### DISCUSSION TITLE
API reference

### DISCUSSION THREAD

**Discussion by plex3r** [fabXlabs/fabX#466] [Q&A]
Hi,

Is there any documentation about the API you can share for example if we wanted to hook up our own design of machine interface what calls can we make to the backend?

Thanks

**Comment by barnyrelph** [comment]
I'm with @plex3r at the same Hackspace - thanks for sharing this project - it looks to pretty neatly line up with what we're trying to do. 

I'm not a Kotlin native speaker, but do work with Java/Hibernate/Docker, so getting the app up and running was pretty straightforward. 

I'm not familiar with the frameworks in use, but I wondered if they generate swagger/OpenAPI or other interface descriptions? 

Also, looking at the Device API from the test classes and the fields persisted, there's clearly the notion of a device having a URL, but I'm not quite sure what the interactions between the device and server would look like? Possibly a sequence diagram or similar would speak a thousand words. Thanks for making this project public and for any support you might be able to give. 

Barny

**Comment by sschmitt98** [comment]
> Is there any documentation about the API you can share for example if we wanted to hook up our own design of machine interface what calls can we make to the backend?

Unfortunately nothing written to be human-readable at the moment, but I like the idea of having it.

In general, the _Device_ (ESP 32) opens a websocket connection to the backend. The device authenticates via basic auth with the username being the device mac address. The password is a secret we generate on first start of the firmware. With the M5 stack we show it on the display as a QR code but it just has to get to the backend somehow.

Json serialised data classes are sent through the websocket. The data classes are defined in [DeviceCommand.kt](https://github.com/fabXlabs/fabX/blob/main/web/src/main/kotlin/cloud/fabX/fabXaccess/device/ws/DeviceCommand.kt). There are a few kinds of these data classes:
* Subclasses of `DeviceToServerCommand` are commands the device sends to the server / backend such that the backend does something. The backend sends back a subclass of `DeviceResponse`.
    * One subclass is the `GetConfiguration` command. This tells the backend to send a `ConfigurationResponse` (with the same `commandId` value). The configuration includes the name of the device and the attached tools among other less important properties. The `attachedTools` are a map from the "pin" to the `ToolConfigurationResponse` for each of the tools. The "pin" is the pin number of the esp32 (or in our case the i2c port expander) to pull high/low to unlock the tool.
    * Another subclass is the `GetAuthorizedTools` command which is probably the most important one. It contains the credentials from a member's card (a `CardIdentity`) or a phone number (a `PhoneIdentity`) to authenticate the member against the backend. The backend then sends a `AuthorizedToolsResponse`. That contains the ids of the tools that the authenticated member is allowed to use _and_ which are attached to the _Device_.
* Subclasses of `DeviceToServerNotification` are similar to `DeviceToServerCommand` but the device does not expect a response back.
  * At the moment this only includes `ToolUnlockedNotification`, where the device tells the backend which of the tools the member chose to unlock.
* Subclasses of `ServerToDeviceCommand` are commands that the backend sends to the device and expects a response back from the device.
  * With the `RestartDevice` command, the server can instruct the device to restart.
  * With the `UpdateDeviceFirmware` command, the server can instruct the device to upgrade its firmware.

> I'm not familiar with the frameworks in use, but I wondered if they generate swagger/OpenAPI or other interface descriptions?

The "web" related framework is [ktor](https://ktor.io/docs/create-server.html), which has an [IntelliJ plugin](https://www.jetbrains.com/help/idea/ktor.html#openapi) for generating an OpenAPI yaml, but I haven't had time to look into, yet.

As far as I know, OpenAPI does not really allow describing a websocket protocol, so it would only be useful for the RESTful API that the frontend uses.

> Also, looking at the Device API from the test classes and the fields persisted, there's clearly the notion of a device having a URL, but I'm not quite sure what the interactions between the device and server would look like?

All entities (devices, users, tools, ...) have URLs to e.g. get their current json representation, e.g. [in DeviceController for getting a Device](https://github.com/fabXlabs/fabX/blob/04d1b28b98ddde6d0da34e873610b0786e32d10e/web/src/main/kotlin/cloud/fabX/fabXaccess/device/rest/DeviceController.kt#L60C13-L60C27).

The devices all open their websocket connection to the same url, though: `wss://{domain}/api/v1/device/ws`, see [here](https://github.com/fabXlabs/fabX/blob/04d1b28b98ddde6d0da34e873610b0786e32d10e/web/src/main/kotlin/cloud/fabX/fabXaccess/device/ws/DeviceWebsocketController.kt#L55).

> Possibly a sequence diagram or similar would speak a thousand words. 

That's a good idea. I'll try to give a short overview in text, here, anyways:

1. Device boots, connects to WiFi, opens websocket connection to the backend (authenticated by mac and secret through basic auth).
2. Device sends a `GetConfiguration` command to the backend. (sample: `{"type":"cloud.fabX.fabXaccess.device.ws.GetConfiguration","commandId":678,"actualFirmwareVersion":"1.42.2"}`)
3. Backend responds with `ConfigurationResponse` which tells the device, e.g., which name it should display. It also tells the device which tools are attached. The device stores this information in memory. The device now knows that tool with id 123 (actually a uuid) is attached to pin 0 and it has to pull the pin low to unlock it.
4. The device idles until a member holds their card up to the reader. The card contains a unique id (uid) and a secret. The device sends a `GetAuthorizedTools` with the card details to the backend.
5. The backend responds with a `AuthorizedToolsResponse` which contains a list of tool ids that the member is authorised to use. The device presents a list of the tool names which the member is authorised to use.
6. The member picks the tool they want to use.
7. The device sends a `ToolUnlockedNotification` to the backend. This is somewhat optional, it's only used for increasing a tool-usage counter metric, at the moment.
8. The device actually unlocks the tool.
9. Depending on the tool type, the device keeps the tool unlocked until the member removes their card or the device just sends a pulse of configurable length on the pin.

You can also take a look at some [integration tests for this websocket connection](https://github.com/fabXlabs/fabX/blob/main/app/src/test/kotlin/cloud/fabX/fabXaccess/device/DeviceIntegrationTest.kt).

> Thanks for making this project public and for any support you might be able to give.

Thank you for giving taking a look at it and writing some kind and helpful comments!
